### PR TITLE
[Macros] Add swift-plugin-server executable

### PR DESCRIPTION
**Pull request by rintaro**
(depends on https://github.com/apple/swift-syntax/pull/1411)

Add `tools/swift-plugin-server`. This executable is intended to be installed in the toolchain and act as an executable compiler plugin just like other 'macro' plugins.

This plugin server has an optional method `loadPluginLibrary` that dynamically loads dylib plugins.
The compiler has a newly added option `-external-plugin-path`. This option receives a pair of the plugin library search path (just like `-plugin-path`) and the corresponding "plugin server" path, separated by `#`. i.e.

```
  -external-plugin-path
    <plugin library search path>#<plugin server exectuable path>
```

For exmaple, when there's a macro decl:

```swift
  @freestanding(expression)
  macro stringify<T>(T) -> (T, String) =
      #externalMacro(module: "BasicMacro", type: "StringifyMacro")
```

The compiler look for `libBasicMacro.dylib` in `-plugin-path` paths first, if not found, it falls back to `-external-plugin-path` and tries to find `libBasicMacro.dylib` in them. If it's found, the "plugin server" path is launched just like an executable plugin, then `loadPluginLibrary` method is invoked via IPC, which `dlopen` the library path in the plugin server. At the actual macro expansion, the mangled name for `BasicMacro.SinglifyMacro` is used to resolve the macro just like dylib plugins in the compiler.

This is useful for
 * Isolating the plugin process, so the plugin crashes doesn't result the compiler crash
 * Being able to use library plugins linked with other `swift-syntax` versions

rdar://105104850

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/apple/swift-syntax/pull/1411
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/apple/swift-syntax/pull/1411
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/apple/swift-syntax/pull/1411
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/apple/swift-syntax/pull/1411
@swift-ci Please smoke test

**Comment by rintaro** [comment]
https://github.com/apple/swift-syntax/pull/1411
@swift-ci Please smoke test

File: tools/swift-plugin-server/CMakeLists.txt
Comment (bnbarham) [inline tools/swift-plugin-server/CMakeLists.txt:2]:
```suggestion
  # _swiftCSwiftPluginServer is just a C support library for swift-plugin-server
```

**Review by bnbarham — APPROVED** [review]

File: include/swift/AST/SearchPathOptions.h
Comment (rintaro) [reply include/swift/AST/SearchPathOptions.h:388]: I will make a custom struct.

`#` is not a special character as a file name at least in macos, (i.e. you can `touch ##.txt` without errors). I just said it's not common to use `#` in directory name or filename, and there's no special meaning in shell, unlike like ';' or `?`.
I agree it's not safe, but...

File: lib/Driver/ToolChains.cpp
Comment (rintaro) [reply lib/Driver/ToolChains.cpp:240]: Yes we need. Since it's auto generated, I want to do that using `main` after I merge this PR

File: include/swift/AST/SearchPathOptions.h
Comment (bnbarham) [reply include/swift/AST/SearchPathOptions.h:388]: No reason to block on this, # is just a little scary to me because of #1/#2 etc. I can definitely imagine *someone* doing that 😅

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: include/swift/AST/TypeCheckRequests.h
Comment (rintaro) [reply include/swift/AST/TypeCheckRequests.h]: Decided not to use `PointerUnion` because we don't know the alignment of `void *`

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Pull request merged**
